src/search/pdbs/pattern_database_factory.h

```diff
@@ -0,0 +1,213 @@
+
+    /*
+      Computes all abstract operators, builds the match tree (successor
+      generator) and then does a Dijkstra regression search to compute
+      all final h-values (stored in distances). operator_costs can
+      specify individual operator costs for each operator for action
+      cost partitioning. If left empty, default operator costs are used.
+    */
+    void create_pdb(
+        const TaskProxy &task_proxy,
+        const std::vector<int> &operator_costs,
+        bool compute_plan,
+        const std::shared_ptr<utils::RandomNumberGenerator> &rng,
+        bool compute_wildcard_plan);
+
+    /*
+      For a given abstract state (given as index), the according values
+      for each variable in the state are computed and compared with the
+      given pairs of goal variables and values. Returns true iff the
+      state is a goal state.
+    */
+    bool is_goal_state(
+        int state_index,
+        const std::vector<FactPair> &abstract_goals,
+        const VariablesProxy &variables) const;
+
+    /*
+      The given concrete state is used to calculate the index of the
+      according abstract state. This is only used for table lookup
+      (distances) during search.
+    */
+    int hash_index(const std::vector<int> &state) const;
+public:
+    /*
+      Important: It is assumed that the passed pattern is
+      sorted, contains no duplicates and is small enough so that the
+      number of abstract states is below numeric_limits<int>::max()
+      Parameters:
+       operator_costs: Can specify individual operator costs for each
+       operator. This is useful for action cost partitioning. If left
+       empty, default operator costs are used.
+       compute_plan: if true, compute an optimal plan when computing
+       distances of the PDB. This requires a RNG object passed via rng.
+       compute_wildcard_plan: when computing a plan (see compute_plan), compute
+       a wildcard plan, i.e., a sequence of parallel operators inducing an
+       optimal plan. Otherwise, compute a simple plan (a sequence of operators).
+    */
+    PatternDatabaseFactory(
+        const TaskProxy &task_proxy,
+        const Pattern &pattern,
+        const std::vector<int> &operator_costs = std::vector<int>(),
+        bool compute_plan = false,
+        const std::shared_ptr<utils::RandomNumberGenerator> &rng = nullptr,
+        bool compute_wildcard_plan = false);
+    ~PatternDatabaseFactory() = default;
+
+    std::shared_ptr<PatternDatabase> extract_pdb();
+
+    // Returns the pattern (i.e. all variables used) of the PDB
+    const Pattern &get_pattern() const {
+        return pattern;
+    }
+
+    // Returns the size (number of abstract states) of the PDB
+    int get_size() const {
+        return num_states;
+    }
+
+    std::vector<std::vector<OperatorID>> && extract_wildcard_plan() {
+        return std::move(wildcard_plan);
+    };
+};
+
+extern std::shared_ptr<PatternDatabase> compute_pdb(
```

Comment:
This function caught my eye as potentially not being good style. I'm not sure if this is the case, and if yes how to best improve it.

Reply:
I think the next step (in a new issue) would be to get rid of the factory class and compute the PDB in a functional way. This makes it possibly simpler to group some of the computed data structures into new classes, such as `pattern`, `num_states`, and `hash_multipliers` into a function `Projection` used for perfect hashing.